### PR TITLE
Multipart form support

### PR DESCRIPTION
**Pull request by frett**
This adds support for submitting multipart/form-data POST requests to an API.

The current need for this is Stories support for GMA/iShare/TBD, the upload image API only accepts multipart/form-data requests.

### CODE REVIEW

**Comment by frett** [comment]
@ryancarlson

File: gto-support-api/src/main/java/org/ccci/gto/android/common/api/MultipartFormOutputWriter.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+final class MultipartFormOutputWriter implements Closeable {
+    private static final String PREFIX = "--";
+    private static final String NEWLINE = "\r\n";
+
+    @NonNull
+    private final DataOutputStream mOut;
+    @NonNull
+    private final String mBoundary;
+
+    MultipartFormOutputWriter(@NonNull final OutputStream os, @NonNull final String boundary) {
```
Comment (ryancarlson) [inline gto-support-api/src/main/java/org/ccci/gto/android/common/api/MultipartFormOutputWriter.java]: Is this intentionally package private?

File: gto-support-api/src/main/java/org/ccci/gto/android/common/api/MultipartFormOutputWriter.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+    private void writeFileParameter(@NonNull final Parameter param) throws IOException {
+        // write boundary
+        mOut.writeBytes(PREFIX);
+        mOut.writeBytes(mBoundary);
+        mOut.writeBytes(NEWLINE);
+
+        // write content header
+        final String fileName =
+                param.mFileName != null ? param.mFileName : param.mFile != null ? param.mFile.getName() : "";
```
Comment (ryancarlson) [inline gto-support-api/src/main/java/org/ccci/gto/android/common/api/MultipartFormOutputWriter.java]: I think parens might help readability here.
````java
final String fileName =
                 ( param.mFileName != null ? param.mFileName : ( param.mFile != null ? param.mFile.getName() : "" ) );
````
or
````java
final String fileName =
                 param.mFileName != null ? param.mFileName : ( param.mFile != null ? param.mFile.getName() : "" );
````

What do you think?

Comment (frett) [inline gto-support-api/src/main/java/org/ccci/gto/android/common/api/MultipartFormOutputWriter.java]: I'll rework it so it's more clear

File: gto-support-api/src/main/java/org/ccci/gto/android/common/api/MultipartFormOutputWriter.java (hunk 1 above)
Comment (frett) [inline gto-support-api/src/main/java/org/ccci/gto/android/common/api/MultipartFormOutputWriter.java]: yes, I don't want this class to be part of the public API

Comment (ryancarlson) [inline gto-support-api/src/main/java/org/ccci/gto/android/common/api/MultipartFormOutputWriter.java]: I guess it's b/c it's only used in a class in this same package.  I tend to create constructors public unless there's some reason not to..  but I think this is totally fine.

Comment (ryancarlson) [inline gto-support-api/src/main/java/org/ccci/gto/android/common/api/MultipartFormOutputWriter.java]: Didn't see your comment when i posted my follow up.... :+1:

**Comment by ryancarlson** [comment]
Overall looks good :+1:

File: gto-support-api/src/main/java/org/ccci/gto/android/common/api/MultipartFormOutputWriter.java (hunk 2 above)
Comment (frett) [inline gto-support-api/src/main/java/org/ccci/gto/android/common/api/MultipartFormOutputWriter.java]: I really wish I could use some of perl's syntactic sugar here

`final String fileName = param.mFileName // param.mFile.getName() // "";`

**Pull request merged**